File: android/app/src/main/java/com/aifencingcoach/runtime/ActivityGatekeeper.kt
package com.aifencingcoach.runtime

import kotlin.math.sqrt

class ActivityGatekeeper(
    private val fps: Int = 30,
    private val activeKneeAngleDeg: Float = 176f,
    private val idleKneeAngleDeg: Float = 178f,
    private val motionThresholdNorm: Float = 0.005f
) {
    var state: String = StateIdle
        private set

    private var frameCount = 0
    private var activeTriggerCount = 0
    private var activeTriggerThreshold = 5
    private var idleTriggerCount = 0
    private var idleTriggerThreshold = 2 * fps
    private var lastPelvisCenter: Point2? = null

    var lastReasons: Map<String, Any?> = emptyMap()
        private set

    fun reset() {
        state = StateIdle
        frameCount = 0
        activeTriggerCount = 0
        idleTriggerCount = 0
        lastPelvisCenter = null
        lastReasons = emptyMap()
    }

    fun shouldExtractPose(): Boolean {
        frameCount += 1
        if (state == StateIdle) {
            val skipRate = maxOf(1, fps / 5)
            return frameCount % skipRate == 1
        }
        return true
    }

    fun update(
        targetSkeleton: Skeleton?,
        opponentSkeleton: Skeleton?,
        frameWidth: Int,
        targetSide: TargetSide
    ): Boolean {
        if (targetSkeleton == null) {
            if (state == StateActive) {
                idleTriggerCount += 1
                if (idleTriggerCount >= idleTriggerThreshold) {
                    state = StateIdle
                    idleTriggerCount = 0
                }
            } else if (state == StateChecking) {
                state = StateIdle
                activeTriggerCount = 0
            }
            lastReasons = mapOf(
                "has_target" to false,
                "state" to state,
                "reason" to "missing_target"
            )
            return state == StateActive
        }

        val kneeAngle = frontKneeAngle(targetSkeleton, targetSide) ?: 180f
        val shoulderWidth = shoulderWidth(targetSkeleton)
        val isTurnedBack = (shoulderWidth ?: frameWidth.toFloat()) < frameWidth * 0.05f

        val pelvisCenter = FencingGeometry.pelvisCenter(targetSkeleton)
        var pelvisMotion = 0f
        if (pelvisCenter != null && lastPelvisCenter != null) {
            val dx = pelvisCenter.x - lastPelvisCenter!!.x
            val dy = pelvisCenter.y - lastPelvisCenter!!.y
            pelvisMotion = sqrt(dx * dx + dy * dy) / frameWidth.toFloat()
        }

        val moving = lastPelvisCenter == null || pelvisMotion >= motionThresholdNorm
        if (pelvisCenter != null) {
            lastPelvisCenter = pelvisCenter
        }

        val enGardePosture = kneeAngle < activeKneeAngleDeg
        val enGarde = enGardePosture && (moving || state != StateIdle || activeTriggerCount > 0)
        val standingUp = kneeAngle > idleKneeAngleDeg
        val stopCondition = standingUp || isTurnedBack

        if (state == StateIdle) {
            if (enGarde) {
                state = StateChecking
                activeTriggerCount = 1
            }
        } else if (state == StateChecking) {
            if (enGarde) {
                activeTriggerCount += 1
                if (activeTriggerCount >= activeTriggerThreshold) {
                    state = StateActive
                    idleTriggerCount = 0
                }
            } else {
                state = StateIdle
                activeTriggerCount = 0
            }
        } else if (state == StateActive) {
            if (stopCondition) {
                idleTriggerCount += 1
                if (idleTriggerCount >= idleTriggerThreshold) {
                    state = StateIdle
                    idleTriggerCount = 0
                }
            } else {
                idleTriggerCount = 0
            }
        }

        lastReasons = mapOf(
            "has_target" to true,
            "state" to state,
            "knee_angle" to kneeAngle,
            "en_garde" to enGarde,
            "en_garde_posture" to enGardePosture,
            "standing_up" to standingUp,
            "turned_back" to isTurnedBack,
            "moving" to moving
        )
        return state == StateActive
    }

    private fun frontKneeAngle(skeleton: Skeleton, targetSide: TargetSide): Float? {
        val limbs = FencingGeometry.frontLimbs.getValue(targetSide)
        val hip = skeleton[limbs.hip] ?: return null
        val knee = skeleton[limbs.knee] ?: return null
        val ankle = skeleton[limbs.ankle] ?: return null
        return FencingGeometry.angle(hip, knee, ankle)
    }

    private fun shoulderWidth(skeleton: Skeleton): Float? {
        val left = skeleton["left_shoulder"]
        val right = skeleton["right_shoulder"]
        return if (left != null && right != null) left.distanceTo(right) else null
    }

    companion object {
        const val StateIdle = "IDLE"
        const val StateChecking = "CHECKING"
        const val StateActive = "ACTIVE"
    }
}

File: backend/inference/activity_gatekeeper.py
"""Activity state machine for gating FenceNet inference."""

from __future__ import annotations

import logging
from typing import Any, Dict, Optional

import numpy as np

logger = logging.getLogger(__name__)


class ActivityGatekeeper:
    """Track whether the target fencer is actively fencing."""

    STATE_IDLE = "IDLE"
    STATE_CHECKING = "CHECKING"
    STATE_ACTIVE = "ACTIVE"

    def __init__(
        self,
        fps: int = 30,
        active_knee_angle_deg: float = 170.0,
        idle_knee_angle_deg: float = 174.0,
        motion_threshold_px: float = 1.5,
    ):
        self.fps = int(fps)
        self.state = self.STATE_IDLE
        self.frame_count = 0
        self.active_knee_angle_deg = float(active_knee_angle_deg)
        self.idle_knee_angle_deg = float(idle_knee_angle_deg)
        self.motion_threshold_px = float(motion_threshold_px)

        self.active_trigger_count = 0
        self.active_trigger_threshold = 5
        self.idle_trigger_count = 0
        self.idle_trigger_threshold = 2 * self.fps

        self.last_pelvis_center: Optional[np.ndarray] = None
        self.last_reasons: Dict[str, Any] = {}

    def should_extract_pose(self) -> bool:
        """Run pose at low FPS while idle and full FPS while checking/active."""
        self.frame_count += 1
        if self.state == self.STATE_IDLE:
            skip_rate = max(1, self.fps // 5)
            return (self.frame_count % skip_rate) == 1
        return True

    def _get_knee_angle(
        self,
        skeleton: Dict[str, Any],
        target_side: str,
    ) -> Optional[float]:
        from .heuristics_engine import FRONT_LIMBS, _get_joint, calc_angle

        limbs = FRONT_LIMBS.get(target_side, FRONT_LIMBS["left"])
        hip = _get_joint(skeleton, limbs["hip"])
        knee = _get_joint(skeleton, limbs["knee"])
        ankle = _get_joint(skeleton, limbs["ankle"])
        if hip is None or knee is None or ankle is None:
            return None
        return calc_angle(hip, knee, ankle)

    def _get_shoulder_width(self, skeleton: Dict[str, Any]) -> Optional[float]:
        from .heuristics_engine import _get_joint

        ls = _get_joint(skeleton, "left_shoulder")
        rs = _get_joint(skeleton, "right_shoulder")
        if ls is None or rs is None:
            return 100.0
        return float(np.linalg.norm(ls - rs))

    def _check_fencer_distance(
        self,
        target_skeleton: Dict[str, Any],
        opponent_skeleton: Optional[Dict[str, Any]],
        frame_width: int,
    ) -> bool:
        if opponent_skeleton is None:
            return False

        from .heuristics_engine import _pelvis_center

        target_center = _pelvis_center(target_skeleton)
        opponent_center = _pelvis_center(opponent_skeleton)
        if target_center is None or opponent_center is None:
            return False

        distance = abs(target_center[0] - opponent_center[0])
        return bool(distance > (frame_width * 0.6))

    def update(
        self,
        target_skeleton: Optional[Dict[str, Any]],
        opponent_skeleton: Optional[Dict[str, Any]],
        frame_width: int,
        target_side: str,
    ) -> bool:
        """Update the state machine and return whether the frame is active."""
        if target_skeleton is None:
            if self.state == self.STATE_ACTIVE:
                self.idle_trigger_count += 1
                if self.idle_trigger_count >= self.idle_trigger_threshold:
                    self.state = self.STATE_IDLE
                    self.idle_trigger_count = 0
            elif self.state == self.STATE_CHECKING:
                self.state = self.STATE_IDLE
                self.active_trigger_count = 0

            self.last_reasons = {
                "has_target": False,
                "state": self.state,
                "reason": "missing_target",
                "active_trigger_count": self.active_trigger_count,
                "idle_trigger_count": self.idle_trigger_count,
            }
            return self.state == self.STATE_ACTIVE

        knee_angle = self._get_knee_angle(target_skeleton, target_side)
        if knee_angle is None:
            knee_angle = 180.0

        shoulder_width = self._get_shoulder_width(target_skeleton)
        is_turned_back = shoulder_width < (frame_width * 0.05)
        too_far = self._check_fencer_distance(
            target_skeleton,
            opponent_skeleton,
            frame_width,
        )

        from .heuristics_engine import _pelvis_center

        pelvis_center = _pelvis_center(target_skeleton)
        pelvis_motion = 0.0
        if pelvis_center is not None and self.last_pelvis_center is not None:
            pelvis_motion = float(np.linalg.norm(pelvis_center - self.last_pelvis_center))
        moving = self.last_pelvis_center is None or pelvis_motion >= self.motion_threshold_px
        if pelvis_center is not None:
            self.last_pelvis_center = pelvis_center

        en_garde_posture = knee_angle < self.active_knee_angle_deg
        en_garde = en_garde_posture and (
            moving
            or self.state != self.STATE_IDLE
            or self.active_trigger_count > 0
        )
        standing_up = knee_angle > self.idle_knee_angle_deg
        stop_condition = standing_up or is_turned_back or too_far

        if self.state == self.STATE_IDLE:
            if en_garde:
                self.state = self.STATE_CHECKING
                self.active_trigger_count = 1
        elif self.state == self.STATE_CHECKING:
            if en_garde:
                self.active_trigger_count += 1
                if self.active_trigger_count >= self.active_trigger_threshold:
                    self.state = self.STATE_ACTIVE
                    self.idle_trigger_count = 0
            else:
                self.state = self.STATE_IDLE
                self.active_trigger_count = 0
        elif self.state == self.STATE_ACTIVE:
            if stop_condition:
                self.idle_trigger_count += 1
                if self.idle_trigger_count >= self.idle_trigger_threshold:
                    self.state = self.STATE_IDLE
                    self.idle_trigger_count = 0
            else:
                self.idle_trigger_count = 0

        self.last_reasons = {
            "has_target": True,
            "state": self.state,
            "knee_angle": float(knee_angle),
            "active_knee_angle_deg": self.active_knee_angle_deg,
            "idle_knee_angle_deg": self.idle_knee_angle_deg,
            "en_garde": bool(en_garde),
            "en_garde_posture": bool(en_garde_posture),
            "standing_up": bool(standing_up),
            "shoulder_width": float(shoulder_width) if shoulder_width is not None else None,
            "turned_back": bool(is_turned_back),
            "too_far": bool(too_far),
            "pelvis_motion": float(pelvis_motion),
            "moving": bool(moving),
            "active_trigger_count": self.active_trigger_count,
            "idle_trigger_count": self.idle_trigger_count,
        }
        return self.state == self.STATE_ACTIVE
